scylla-cql/src/types/deserialize/mod.rs

```diff
@@ -0,0 +1,307 @@
+//!          if let ColumnType::Blob = typ {
+//!              return Ok(());
+//!          }
+//!          Err(ParseError::BadIncomingData("Expected bytes".to_string()))
+//!      }
+//!
+//!      fn deserialize(
+//!          _typ: &'frame ColumnType,
+//!          v: Option<FrameSlice<'frame>>,
+//!      ) -> Result<Self, ParseError> {
+//!          v.ok_or_else(|| {
+//!             ParseError::BadIncomingData("Expected non-null value".to_string())
+//!          })
+//!          .map(|v| Self(v.as_slice()))
+//!      }
+//! }
+//! ```
+//!
+//! ## Reference-counted types (`DeserializeCql` only)
+//!
+//! Internally, the driver uses the `bytes::Bytes` type to keep the contents
+//! of the serialized response. It supports creating derived `Bytes` objects
+//! which point to a subslice but keep the whole, original `Bytes` object alive.
+//!
+//! During deserialization, a type can obtain a `Bytes` subslice that point
+//! to the serialized value. This approach combined advantages of the previous
+//! two approaches - creating a derived `Bytes` object can be cheaper than
+//! allocation and a copy (it supports `Arc`-like semantics) and the `Bytes`
+//! type is not constrained by a lifetime. However, you should be aware that
+//! the subslice will keep the whole `Bytes` object that holds the frame alive.
+//! It is not recommended to use this approach for long-living objects because
+//! it can introduce space leaks.
+//!
+//! Example:
+//!
+//! ```rust
+//! # use scylla_cql::types::deserialize::{value::DeserializeCql, FrameSlice};
+//! # use scylla_cql::frame::response::result::ColumnType;
+//! # use scylla_cql::frame::frame_errors::ParseError;
+//! # use bytes::Bytes;
+//! struct MyBytes(Bytes);
+//! impl<'frame> DeserializeCql<'frame> for MyBytes {
+//!     fn type_check(typ: &ColumnType) -> Result<(), ParseError> {
+//!          if let ColumnType::Blob = typ {
+//!              return Ok(());
+//!          }
+//!          Err(ParseError::BadIncomingData("Expected bytes".to_string()))
+//!      }
+//!
+//!      fn deserialize(
+//!          _typ: &'frame ColumnType,
+//!          v: Option<FrameSlice<'frame>>,
+//!      ) -> Result<Self, ParseError> {
+//!          v.ok_or_else(|| {
+//!              ParseError::BadIncomingData("Expected non-null value".to_string())
+//!          })
+//!          .map(|v| Self(v.to_bytes()))
+//!      }
+//! }
+//! ```
+
+pub mod row;
+pub mod value;
+
+use bytes::Bytes;
+
+use crate::frame::frame_errors::ParseError;
+use crate::frame::types;
+
+/// A reference to a part of the frame.
+#[derive(Clone, Copy, Debug)]
+pub struct FrameSlice<'frame> {
+    mem: &'frame [u8],
+    frame: &'frame Bytes,
+}
+
+static EMPTY_BYTES: Bytes = Bytes::new();
+
+impl<'frame> FrameSlice<'frame> {
+    /// Creates a new FrameSlice from a reference of a Bytes object.
+    ///
+    /// This method is exposed to allow writing deserialization tests
+    /// for custom types.
+    #[inline]
```

Comment:
If a function is marked `#[inline]`, the compiler is permitted to inline it if it is used from a different crate. If it's not marked like that then it won't be inlined unless LTO is enabled or the function is generic.

I added the `#[inline]` annotation for small functions that will most likely be called by user code. The annotation is still a hint and not a requirement for the compiler, but e.g. in case of trivial getters inlining them should be a win pretty much always.

Reply:
I wasn't aware that `#[inline]` enables inlining through crate boundary without LTO enabled. GTK.